Lib/test/test_asyncio/test_proactor_events.py

```diff
@@ -489,6 +407,191 @@ def test_dont_pause_writing(self):
         self.assertFalse(self.protocol.pause_writing.called)
 
 
+class ProactorSocketTransportWithProtocolTests(test_utils.TestCase, ProactorSocketTransportTestsBase):
+
+    def make_test_protocol(self):
+        self.protocol = test_utils.make_test_protocol(asyncio.Protocol)
+
+    def setUp(self):
+        super().setUp()
+        self.setup_globals()
+
+    def test_ctor(self):
+        fut = self.loop.create_future()
+        tr = self.socket_transport(waiter=fut)
+        test_utils.run_briefly(self.loop)
+        self.assertIsNone(fut.result())
+        self.protocol.connection_made(tr)
+        self.proactor.recv.assert_called_with(self.sock, self.buffer_size)
+
+    def test_loop_reading(self):
+        tr = self.socket_transport()
+        tr._loop_reading()
+        self.loop._proactor.recv.assert_called_with(self.sock, self.buffer_size)
+        self.assertFalse(self.protocol.data_received.called)
+        self.assertFalse(self.protocol.eof_received.called)
+
+    def test_loop_reading_data(self):
+        buf = b'data'
+        res = self.loop.create_future()
+        res.set_result(buf)
+
+        tr = self.socket_transport()
+        tr._read_fut = res
+        tr._loop_reading(res)
+        self.loop._proactor.recv.assert_called_with(self.sock, self.buffer_size)
+        self.protocol.data_received.assert_called_with(buf)
+
+    def test_pause_resume_reading(self):
+        tr = self.socket_transport()
+        futures = []
+        msgs = [b'data1', b'data2', b'data3', b'data4', b'data5', b'']
+        for msg in msgs:
+            f = self.loop.create_future()
+            f.set_result(msg)
+            futures.append(f)
+
+        self.loop._proactor.recv.side_effect = futures
+        self.loop._run_once()
+        self.assertFalse(tr._paused)
+        self.assertTrue(tr.is_reading())
+
+        for msg in msgs[:2]:
+            self.loop._run_once()
+            self.protocol.data_received.assert_called_with(msg)
```

Comment:
before the `_run_once()` would not it be better to assert that the `msg` is not in the `_mock_call_args_list` of the mock object ?

and same below for the others `data_received.assert_called_with` following a `_run_once()`.

Reply:
I don't understand which mock object you are talking about, and I don't see any other test reading `_mock_call_args_list`.

Can you elaborate?

Reply:
something like:

```python
for msg in msgs[:2]:
            self.assertFalse(unittest.mock.call(msg) in self.protocol.data_received.mock_calls)
            self.loop._run_once()
            self.protocol.data_received.assert_called_with(msg)
```

to ensure that each `msg` is effectively processed during the _run_once() call.

Reply:
Oh I see, but protocol's `buffer_updated` (data_received is only in a regular stream object, not buffered) is not a mock, it's an actual function, so no `mock_calls` field.

Reply:
ah ok :/